pkgs/development/compilers/llvm/common/libcxx/default.nix

```
@@ -123,9 +129,28 @@ stdenv.mkDerivation (rec {
   # libc++.so is a linker script which expands to multiple libraries,
   # libc++.so.1 and libc++abi.so or the external cxxabi. ld-wrapper doesn't
   # support linker scripts so the external cxxabi needs to be symlinked in
-  postInstall = lib.optionalString (cxxabi != null) ''
+  postInstall = ''
+    libcxxabi=$out/lib/libc++abi.a
+  '' + lib.optionalString (cxxabi != null) ''
     lndir ${lib.getDev cxxabi}/include $dev/include/c++/v1
     lndir ${lib.getLib cxxabi}/lib $out/lib
+    libcxxabi=$out/lib/lib${cxxabi.libName}.a
+  '' + lib.optionalString (!lib.versionAtLeast release_version "16") ''
+    # Note: llvm >= 16 implements this with LIBCXX_STATICALLY_LINK_ABI_IN_STATIC_LIBRARY=On.
+    if [[ -f $out/lib/libc++.a && -e $libcxxabi ]]; then
+      scratch="$(mktemp -d)"
+      mv $out/lib/libc++.a $scratch/libcxx.a
+      cp -d $libcxxabi $scratch/libcxxabi.a
+      $AR -M <<MRI
+      create libcxx.a
+      addlib $scratch/libcxx.a
+      addlib $scratch/libcxxabi.a
+      save
+      end
+    MRI
+      mv libcxx.a $out/lib/libc++.a
+      rm -rf $scratch
```

this can be simplified to:
```suggestion
     $AR -M <<MRI
        create $out/lib/libc++.a
        addlib $out/lib/libc++.a
        addlib $libcxxabi
        save
        end
    MRI
```

`ar` in newer versions of binutils can handle `+`.  (i initially was testing with ar from binutils 2.38 which can't handle `+`).

also. note that gnu binutils adds objects to the archive in LIFO order, so when you look at the result `ar t result/lib/libc++.a` the objects will be reversed. this can be fixed, if desired, by adding an additional mri script.
```bash
      $AR -M <<MRI
        create $out/lib/libc++.a
        addlib $out/lib/libc++.a
        save
        end
    MRI
```

this is what is done in the diff embedded in  https://github.com/NixOS/nixpkgs/pull/307184#issuecomment-2080457122